### ISSUE TITLE
How does limbo affect tabs?

### ISSUE THREAD

**Issue by madebysid** [prisma/studio#86]
Currently, these are two separate entities that do not affect each other.

Cases I'm thinking of:

1. Does creating a new record make the current tab dirty?
2. If you create a new row in a preview tab, does it become sticky?

There might be a few other cases that we might want to discuss.

**Comment by kuldar** [comment]
Yes, I think limbo state should turn the tab (both preview and sticky) into a dirty tab.

**Comment by madebysid** [comment]
The thing is, when you’re adding/updating a record, you’re not really changing what is saved (the model/saved tab). You don't save records in the sidebar, you save a script that fetches these records.

So while I understand the idea of making a tab dirty, having limbo as part of this raises other bigger problems: What happens if you close this tab, saving changes? Let's say we somehow commit ONLY changes that you made in this tab. First: this might not always be possible. Assuming it is, now every other tab's dirty state might need changing. Additionally, other non-committed changes might become invalid now.

Concrete examples:

You have Users and Posts, both open in two separate tabs.
Terminology:
action = anything you do that changes the limbo state
committing = flush action to database.

1. You create a user, you create a post. You close the User tab, saving changes. This is straight-forward. Each limbo action was atomic and can be committed. ✅
2. You create a user, and connect it to a post. At this time, both tabs are dirty. Now, you try to close the Post tab, saving changes. Here's a problem. You cannot commit the Post without also committing the User (because the user you're trying to connect this post to does not exist yet). ❌
3. You have two saved tabs, they're both IDENTICAL. You open them both in different tabs. You change the `name` of User1 in tab1 to 'ABC', you change the `name` of User1 in tab2 to 'ABC' as well. Both tabs are dirty. You close tab1, saving changes. Now suddenly, saving tab1 has made tab2 non-dirty as well. ❌

All these examples get even more complex if you add a third model into the mix.

My point, again, is that `Saved Tabs` are really `Saved Scripts`, so the dirty status should reflect changes to scripts, not what these scripts do.

**Comment by kuldar** [comment]
If I'm understanding the problem correctly, the complexity comes from trying to access "uncommited" data changes from another tab?

I would see the tab limbo states to be more self contained. So when you have an "uncommited" new user in `tab 1`, you can't connect a post to it in `tab 2`, before you actually commit the changes to the database in the first tab.

**Comment by madebysid** [comment]
Ah so the limbo state is local to a tab, rather than global as it is right now. That could work.

So then I'd imagine if you have the same script open in two tabs and change one of them, one tab would see the limbo state, but the other won't (until you commit). That makes more sense than a global limbo state!

Also on a side note, we should warn people that saving a tab with limbo means committing to your DB as well, but that should be trivial.

**Comment by kuldar** [comment]
I think the "query changes" and "data changes" should probably have a separate confirmation.

1. If you close a tab with no query changes but some pending limbo state, you'll be asked to commit or discard the data changes.
2. When you close a tab that has both query changes and pending limbo state, you'll be first asked about commiting or discarding the data changes, and then asked to save or discard the query changes.

Does this seem like a sensible UX?

Because I can imagine situations where you might want to open a saved query, adjust the query a bit, make some changes to the data, and then wanting to save the data changes to db, but ignore the query changes for the saved query.

**Comment by madebysid** [comment]
That's valid, I can see that happening too!
